File: [Objective-C]/Commons/Geometry.m
//
//  Geometry.m
//  Karma and destiny
//
//  Created by Олег on 16.11.2017.
//  Copyright © 2017 Олег. All rights reserved.
//

#import "Geometry.h"

@implementation Geometry
{
    CGRect _rect;
}

+(Geometry *(^)(CGSize))size
{
    return ^id(CGSize size) {
        Geometry *g = [[Geometry alloc] init];
        g->_rect = CGRectMake(0, 0, size.width, size.height);
        return g;
    };
}
+(Geometry *(^)(CGRect))rect
{
    return ^id(CGRect rect) {
        Geometry *g = [[Geometry alloc] init];
        g->_rect = rect;
        return g;
    };
}

+(Geometry *(^)(CGPoint))point {
    return ^id(CGPoint point) {
        Geometry *g = [[Geometry alloc] init];
        g->_rect = CGRectMake(point.x, point.y, 0, 0);
        return g;
    };
}

-(Geometry *(^)(CGSize))fill {
    return ^id(CGSize sizeToFill) {
        CGSize size = _rect.size;
        float ratio = size.width / size.height;
        CGSize szFill = ratio <  sizeToFill.width / sizeToFill.height ? CGSizeMake(sizeToFill.width, sizeToFill.width / ratio) : CGSizeMake(sizeToFill.height * ratio, sizeToFill.height);
        CGPoint pntFill = CGPointMake(sizeToFill.width/2.-szFill.width/2., sizeToFill.height/2.-szFill.height/2.);
        _rect =  CGRectMake(pntFill.x, pntFill.y, szFill.width, szFill.height);
        return self;
    };
}

-(Geometry *(^)(CGSize))fit
{
    return ^id(CGSize sizeToFit) {
        CGSize size = _rect.size;
        float ratio = size.width / size.height;
        CGSize szFit = (ratio > sizeToFit.width / sizeToFit.height) ? CGSizeMake(sizeToFit.width, sizeToFit.width / ratio) : CGSizeMake(sizeToFit.height * ratio, sizeToFit.height);
        CGPoint pntFit = CGPointMake(sizeToFit.width/2.-szFit.width/2., sizeToFit.height/2.-szFit.height/2.);
        _rect = CGRectMake(pntFit.x, pntFit.y, szFit.width, szFit.height);
        return self;
    };
}
-(Geometry *(^)(CGFloat))scale
{
    return ^id(CGFloat scale) {
        _rect = CGRectMake(_rect.origin.x*scale, _rect.origin.y*scale, _rect.size.width*scale, _rect.size.height*scale);
        return self;
    };
}

-(Geometry *(^)(CGFloat))shift
{
    return ^id(CGFloat shift) {
        _rect = CGRectMake(_rect.origin.x+shift, _rect.origin.y+shift, _rect.size.width-shift*2., _rect.size.height-shift*2.);;
        return self;
    };
}
-(Geometry *(^)(CGPoint))move
{
    return ^id(CGPoint move) {
        _rect.origin.x += move.x;
        _rect.origin.y += move.y;
        return self;
    };
}
-(Geometry *)roundf{
    _rect = CGRectMake(round(_rect.origin.x), round(_rect.origin.y), round(_rect.size.width), round(_rect.size.height));
    return self;
}

-(CGRect)rect
{
    return _rect;
}
-(CGPoint)point
{
    return _rect.origin;
}
-(CGSize)size
{
    return _rect.size;
}
-(CGPoint)center
{
    return CGPointMake(_rect.origin.x + _rect.size.width/2., _rect.origin.y + _rect.size.height/2.);
}

-(CGFloat)x
{
    return _rect.origin.x;
}
-(CGFloat)y
{
    return _rect.origin.y;
}
-(CGFloat)width
{
    return _rect.size.width;
}
-(CGFloat)height
{
    return _rect.size.height;
}
@end

File: UICostelique/UICostelique/[Swift]/Geometry.swift
//
//  Geometry.swift
//  Action Camera Master
//
//  Created by Олег on 05.05.17.
//  Copyright © 2017 Олег. All rights reserved.
//

import UIKit.UIGeometry
import CoreGraphics.CGGeometry

public extension CGSize {
    init(_ x:CGFloat,_ y:CGFloat) {
        self.init(width: x, height: y)
    }

    func size(withWidth:CGFloat) -> CGSize {
        return CGSize.init(width: withWidth, height:self.height)
    }
    
    func round() -> CGSize {
        return CGSize.init(width: CGFloat(Darwin.round(Double(self.width))), height: CGFloat(Darwin.round(Double(self.height))))
    }
    
    func fill(_ size:CGSize) -> CGSize {
        return fill(width:size.width, height:size.height)
    }
    
    func fill(width:CGFloat, height:CGFloat) -> CGSize {
        if (self.width / width < self.height / height) {
            return CGSize.init(width: width, height: width * self.height / self.width)
        }
        else {
            return CGSize.init(width: height*self.width / self.height, height: height)
        }
    }
    
    func fit(_ size:CGSize) -> CGSize {
        return fit (width: size.width, height: size.height)
    }
    
    func fit(width:CGFloat, height:CGFloat) -> CGSize {
        if (self.width / width > self.height / height) {
            return CGSize.init(width: width, height: width * self.height / self.width)
        }
        else {
            return CGSize.init(width: height * self.width / self.height, height: height)
        }
    }
    
    func fitRect(_ size:CGSize) -> CGRect {
        let fitSize = fit(size)
        return CGRect(x: size.width/2 - fitSize.width/2, y: size.height/2 - fitSize.height/2, width: fitSize.width, height: fitSize.height)
    }

    func fitRect(_ rect:CGRect) -> CGRect {
        let size = rect.size
        let fitSize = fit(size)
        var out = CGRect(x: size.width/2 - fitSize.width/2, y: size.height/2 - fitSize.height/2, width: fitSize.width, height: fitSize.height)
        out.origin.x += rect.origin.x
        out.origin.y += rect.origin.y
        return out
    }
    
    
    func fillRect(_ size:CGSize) -> CGRect {
        let fillSize = fill(size)
        return CGRect(x: size.width/2 - fillSize.width/2, y: size.height/2 - fillSize.height/2, width: fillSize.width, height: fillSize.height)
    }
    
    func point() -> CGPoint {
        return CGPoint(x: self.width, y: self.height)
    }
    
    func scale(_ scale: CGFloat) -> CGSize {
        return CGSize(width: self.width * scale, height: self.height * scale)
    }


    //Math
    static func * (left: CGSize, right:CGFloat) -> CGSize {
        return CGSize(width:left.width * right, height: left.height * right)
    }

    static func / (left: CGSize, right:CGFloat) -> CGSize {
        return CGSize(width:left.width / right, height: left.height / right)
    }

    static func - (left: CGSize, right: CGSize) -> CGSize {
        return CGSize(width: left.width - right.width, height: left.height - right.height)
    }

    static func + (left: CGSize, right: CGSize) -> CGSize {
        return CGSize(width: left.width + right.width, height: left.height + right.height)
    }
}

public extension CGPoint {
    init(_ x:CGFloat,_ y:CGFloat) {
        self.init(x: x, y: y)
    }
    func distance(to point:CGPoint) -> CGFloat {
        let x : Double = Double(point.x - self.x)
        let y : Double = Double(point.y - self.y)
        return CGFloat(sqrt(x * x) + sqrt(y * y))
    }
    func round() -> CGPoint {
        return CGPoint(CGFloat(Darwin.round(Double(x))), CGFloat(Darwin.round(Double(y))))
    }

    //Math
    static func - (left: CGPoint, right: CGPoint) -> CGPoint {
        return CGPoint(x: left.x - right.x, y: left.y - right.y)
    }

    static func + (left: CGPoint, right: CGPoint) -> CGPoint {
        return CGPoint(x: left.x + right.x, y: left.y + right.y)
    }

    static func * (left: CGPoint, right: CGFloat) -> CGPoint {
        return CGPoint(x: left.x * right, y: left.y * right)
    }

    static func / (left: CGPoint, right: CGFloat) -> CGPoint {
        return CGPoint(x: left.x / right, y: left.y / right)
    }
}

public extension CGRect {

    init(_ x:CGFloat,_ y:CGFloat,_ w:CGFloat,_ h:CGFloat) {
        self.init(x: x, y: y, width: w, height: h);
    }
    func shift(_ shift: CGFloat) -> CGRect {
        return CGRect(x: self.origin.x + shift,
                      y: self.origin.y + shift,
                      width: self.size.width - 2 * shift,
                      height: self.size.height - 2 * shift)
    }
    
    func shift(_ insets: UIEdgeInsets) -> CGRect {
        return CGRect(x: self.origin.x + insets.left,
                      y: self.origin.y + insets.bottom,
                      width: self.size.width - insets.left - insets.right,
                      height: self.size.height - insets.top - insets.bottom)
    }
    
    func center () -> CGPoint {
        return CGPoint(x: self.origin.x + self.size.width / 2,
                       y: self.origin.y + self.size.height / 2)
    }
    
    func round() -> CGRect {
        let rx = CGFloat(Darwin.round(Double(self.origin.x)))
        let ry = CGFloat(Darwin.round(Double(self.origin.y)))
        let rwidth = CGFloat(Darwin.round(Double(self.size.width)))
        let rheight = CGFloat(Darwin.round(Double(self.size.height)))
        return CGRect(x:rx, y:ry, width:rwidth, height: rheight)
    }
}




